vsintegration/src/FSharp.Editor/CodeFixes/ChangeEqualsInFieldTypeToColon.fs

```diff
@@ -0,0 +1,57 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+namespace Microsoft.VisualStudio.FSharp.Editor
+
+open System.Collections.Immutable
+
+open Microsoft.CodeAnalysis
+open Microsoft.CodeAnalysis.Text
+open Microsoft.CodeAnalysis.CodeFixes
+
+open FSharp.Compiler.Text
+
+open CancellableTasks
+
+[<ExportCodeFixProvider(FSharpConstants.FSharpLanguageName, Name = CodeFix.ChangeEqualsInFieldTypeToColon)>]
+type internal ChangeEqualsInFieldTypeToColonCodeFixProvider() =
+    inherit CodeFixProvider()
+
+    let isInRecord (document: Document) (range: range) =
+        cancellableTask {
+            let! parseResults = document.GetFSharpParseResultsAsync(nameof ChangeEqualsInFieldTypeToColonCodeFixProvider)
+
+            return parseResults.IsPositionWithinRecordDefinition(range.Start)
+        }
+
+    override _.FixableDiagnosticIds = ImmutableArray.Create("FS0010")
+
+    override this.RegisterCodeFixesAsync context = context.RegisterFsharpFix(this)
```

Comment:
We've been thinking about the same issue with localization in Rider for some time, and I think the best approach we could use is to [add the relevant data into the diagnostics](https://github.com/fsharp/fslang-suggestions/issues/1094), e.g. instances of `FSharpType` for type errors, tokens text/name for `FS0010` where possible and so on. Union cases holding info for specific diagnostic types seems to be the most accepted option from the previous discussions in Slack/Twitter/etc.

Reply:
I think subtypes would be more binary-future-compatible compared to DUs, but otherwise yes, this could be a valid appraoch and also possible to roll incrementally (I think that is a must given the diagnostics existing in the codebase already).

Reply:
Btw. the calls to FSComp.SR.  are going trough resources and take the CurrentUICulture.
This is no VS magic, is all in the compiler code.


Oh and an idea of at least quickly gaining accessed to structured data, even though unnamed ("named" would have to go via dedicated subtypes/cases as written above) - the arguments which now in as format specifiers might be easily&centrally extractable - FsComp.fs/createMessageString has access to it.

But personally, I am now more inclined to do diagnostics subtypes on a as-needed basis when a specific feature demands it.
(incl. my favourit subtype for a diagnostic that would already include a lazy TextChange proposal, since the error/warning emitting code is IMO the one best equipped to produce it)

Reply:
> I think subtypes would be more binary-future-compatible compared to DUs

@T-Gro Could you please elaborate? If we start with more than one case in a DU, then I don't see how adding another case later would break binary compatibility. Each case with fields would be compiled into a nested subtype anyway.

Using DU seems a bit better for pattern matching over the fields in F#: it won't require type check pattern + `as` pattern to access the field later in expression. Checking particular cases might also be optimized by the compiler to just checking an `int` union case tag, compared to always required type checking with explicit subtypes.

Reply:
@auduchinok :

Firstly, the internal representation of a DU can change as the DU grows (e.g. from being a single-value wrapper to a choice later) or changes shape (nullary vs non-nullary cases as well).

The more important issue I see, with the lenses of "using a a newer version of F.C.S without having to recompile" is that you might start with exhaustive matching, but then F.C.S adds a new case -> you suddenly start getting runtime exceptions even though at coding time it was all considered safe.

I also see this as something which would have to play together with custom analyzers, where the "upgrade F.C.S without recompiling" is a likely scenario.

(I admit that in the tooling world it's probably a non-issue)